### PR TITLE
Hint user to update nightly on ICEs produced from outdated nightly

### LINKED ISSUE #118832: hint users to update nightly on ICEs of outdated nightly
From time to time we get reports of regressions that have already been fixed since several days, but the user uses a nightly that is 1-2 weeks old for example.

We could probably make nightly somewhat self-aware and make it:
grab the current date and compare it to its own version string
calculate when `$current_version` was released (iirc the time when the master-> nightly promotion happens can be found on the forge somewhere)  and add maybe 6 to 12 hours as buffer.
If `$current_version.release_timestamp` is older than  $current_time + 24 hours + buffer  and nightly crashes, add a hint to perhaps update to the user but only iff we are on the master/nightly channel!

could look like this:

````
note: it seems this 'rustc 1.76.0-nightly (d86d65bbc 2023-12-10)' is outdated, nightly 2023-12-11 should have been released in the meantime. 
note: please consider running `rustup update nightly` to update the nightly channel and checking if this problem still persists.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please attach the file at `/tmp/im/rustc-ice-2023-12-11T16_34_58-2325463.txt` to your bug report

````

We could also consider hiding/adapting the `we would appreciate a bug report` line in case we detected a potentially outdated nightly.

### PR DESCRIPTION
**Pull request by jieyouxu**
This is a conservative best-effort approach to detect a potentially outdated nightly; it will fallback to the regular ICE-reporting if any of the following cases are true:
- Channel is not nightly
- Version information is not available
- Version date is not parseable as a YYYY-MM-DD or is missing
- System time is at least 36 hours ahead of the user's nightly release datetime.
- Any internal features are used.

Note that I'm not sure how to make a test for this: I tested this manually by `CFG_VER_DATE="2020-02-02" ./x build library --stage 1`, and also changing the channel detection in `rustc_driver_impl` from `Some("nightly")` to `Some("nightly" | "dev")`, and then running `rustc +stage1 test.rs -Ztreat-err-as-bug=1` with a non-existent `test.rs`.

Screenshot 2024-02-27 at 01 12 28

Closes #118832.

### CODE REVIEW

File: compiler/rustc_driver_impl/src/lib.rs [hunk 1]
```diff
@@ -1398,10 +1401,41 @@ fn report_ice(
...
+    use time::ext::NumericalDuration;
+
+    // Try to hint user to update nightly if applicable when reporting an ICE.
+    // Attempt to calculate when current version was released, and add 12 hours
+    // as buffer. If the current version's release timestamp is older than
+    // the system's current time + 24 hours + 12 hours buffer if we're on
+    // nightly.
+    if let Some("nightly") = option_env!("CFG_RELEASE_CHANNEL")
+        && let Some(version) = option_env!("CFG_VERSION")
+        && let Some(ver_date_str) = option_env!("CFG_VER_DATE")
```
Comment (jieyouxu) [inline compiler/rustc_driver_impl/src/lib.rs:1413]: I'm *assuming* these must be set by bootstrap for nightly builds?

**Comment by oli-obk** [comment]
is it useful to recommend updating at all if an internal feature was included?

File: compiler/rustc_driver_impl/src/lib.rs (hunk 1 above)
Comment (oli-obk) [reply compiler/rustc_driver_impl/src/lib.rs:1413]: good question 😆 can do some investigation? If it turns out inconclusive, we'll just land it and check the next nightly 🙃

Comment (jieyouxu) [reply compiler/rustc_driver_impl/src/lib.rs:1413]: > good question 😆 can do some investigation? If it turns out inconclusive, we'll just land it and check the next nightly 🙃

My assumption is based on

https://github.com/rust-lang/rust/blob/53ed660d47feb01055483fe81b628d5ef9705dbd/compiler/rustc_driver_impl/src/lib.rs#L846-L862

and 

https://github.com/rust-lang/rust/blob/53ed660d47feb01055483fe81b628d5ef9705dbd/compiler/rustc_session/src/errors.rs#L39-L49

so I'm fairly certain they are available for nightly compilers.

**Pull request merged**